src/main/java/org/zalando/nakadi/service/timeline/TimelineSyncImpl.java

```diff
@@ -0,0 +1,365 @@
+                final List<Consumer<String>> toNotify;
+                synchronized (consumerListeners) {
+                    toNotify = consumerListeners.containsKey(unlocked) ?
+                            new ArrayList<>(consumerListeners.get(unlocked)) : null;
+                }
+                if (null != toNotify) {
+                    for (final Consumer<String> listener : toNotify) {
+                        try {
+                            listener.accept(unlocked);
+                        } catch (final RuntimeException ex) {
+                            log.error("Failed to notify about event type {} unlock", unlocked, ex);
+                        }
+                    }
+                }
+            }
+            try {
+                zooKeeperHolder.get().setData()
+                        .forPath(toZkPath("/nodes/" + thisId), String.valueOf(change.version).getBytes(CHARSET));
+            } catch (final Exception ex) {
+                log.error("Failed to update node version in zk. Will try to reprocess again", ex);
+                return;
+            }
+            queuedChanges.poll();
+            log.info("Delayed change {} successfully processed", change);
+        }
+    }
+
+    /**
+     * MUST be executed under zk lock. Called when version is changed
+     */
+    private void refreshVersion() {
+        try {
+            log.info("Adding refresh call to delayed changes list");
+            queuedChanges.add(
+                    new DelayedChange(readData("/version", this::versionChanged, Integer::parseInt),
+                            this.zooKeeperHolder.get().getChildren().forPath(toZkPath("/locked_et"))));
+        } catch (final RuntimeException ex) {
+            throw ex;
+        } catch (final Exception ex) {
+            throw new RuntimeException(ex);
+        }
+    }
+
+    private void versionChanged(final WatchedEvent ignore) {
+        refreshVersion();
+    }
+
+    private void runLocked(final Runnable action) {
+        try {
+            Exception releaseException = null;
+            final InterProcessLock lock =
+                    new InterProcessSemaphoreMutex(zooKeeperHolder.get(), ROOT_PATH + "/lock");
+            lock.acquire();
+            try {
+                action.run();
+            } finally {
+                try {
+                    lock.release();
+                } catch (final Exception ex) {
+                    log.warn("Failed to release lock", ex);
+                    releaseException = ex;
+                }
+            }
+            if (null != releaseException) {
+                throw releaseException;
+            }
+        } catch (final RuntimeException ex) {
+            throw ex;
+        } catch (final Exception ex) {
+            throw new RuntimeException(ex);
+        }
+    }
+
+    @Override
+    public Closeable workWithEventType(final String eventType) throws InterruptedException {
+        synchronized (localLock) {
```

Comment:
Yes, but the publishing in this case looks like this: 
1. lock
2. unlock
3. publish
4. lock 
5. unlock

I hope that it will be enough. I can include locking right now to publishing and we can release it to be sure that it's fast enough during very first release.